File: src/python/api/editor.py
from flask import Blueprint, request, Response, jsonify
import pystache

from ..shared import Config
from ..engine.search import mongo_filter, sorter

api_editor = Blueprint("editor", __name__, url_prefix="/api/editor")


@api_editor.route("/", methods=["POST", "PUT", "DELETE"])
def r_editor():
    r = request.json
    db = Config.DB

    if request.method == "POST":
        offset = r.get("offset", 0)
        all_data = sorted(filter(mongo_filter(r["q"]), db.get_all()),
                          key=sorter(r.get("sortBy", "deck"), r.get("desc", False)))
        return jsonify({
            "data": list(map(_editor_entry_post_process, all_data[offset: offset + r.get("limit", 10)])),
            "count": len(all_data)
        })

    elif request.method == "PUT":
        if r["create"]:
            c_id = db.insert_many([r["create"]])[0]
            return jsonify({
                "id": c_id
            })

        if r["update"]:
            db.update(r["id"], r["update"])
        else:
            db.update(r["id"], {r["fieldName"]: r["fieldData"]})

        return Response(status=201)

    elif request.method == "DELETE":
        db.conn.execute("""
        DELETE FROM card
        WHERE id = ?
        """, (r["id"],))

        return Response(status=201)

    return Response(status=404)


@api_editor.route("/findOne", methods=["POST"])
def r_editor_find_one():
    c = tuple(filter(mongo_filter({"id": request.json["id"]}), Config.DB.get_all()))[0]

    if c["front"].startswith("@md5\n"):
        c["front"] = c["tFront"]
        c["back"] = c["tBack"]

    return jsonify(c)


@api_editor.route("/insertMany", methods=["POST"])
def r_editor_insert_many():
    return jsonify(Config.DB.insert_many(request.json["entries"]))


def _editor_entry_post_process(c: dict) -> dict:
    data = c.get("data", dict())
    if c["front"].startswith("@md5\n"):
        c["front"] = pystache.render(c["tFront"], data)
        c["back"] = pystache.render(c["tBack"], data)

    return c

File: src/node/api/editor.ts
import { SearchParser } from "../engine/search";
import { Router } from "express";
import g from "../config";

const router = Router();

router.post("/", (req, res) => {
    const {q, offset, limit} = req.body;
    const db = g.DB!;
    const parser = new SearchParser();

    return res.json(db.parseCond(parser.doParse(q) || {}, {offset, limit,
        fields: [
            "front", "back", "mnemonic", "tag", "srsLevel", "nextReview", "created", "modified", "stat",
            "deck",
            "template", "model", "tFront", "tBack", "css", "js",
            "key", "data",
            // "source", "sH", "sCreated"
        ]
    }));
});

router.put("/", (req, res) => {
    const {id, ids, create, update} = req.body;
    const db = g.DB!;
    if (Array.isArray(create)) {
        const ids = db.insertMany(create);
        return res.json({ids});
    } else if (create) {
        const ids = db.insertMany([create]);
        return res.json({id: ids[0]});
    } else if (ids) {
        return res.json(db.updateMany(ids, update));
    } else {
        return res.json(db.updateMany([id], update));
    }
});

router.delete("/", (req, res) => {
    const {id, ids} = req.body;
    const db = g.DB!;
    if (ids) {
        return res.json(db.deleteMany(ids));
    } else {
        return res.json(db.deleteMany([id]));
    }
})

router.put("/addTags", (req, res) => {
    const {ids, tags} = req.body;
    const db = g.DB!;
    return res.json(db.addTags(ids, tags));
});

router.delete("/editTags", (req, res) => {
    const {ids, tags} = req.body;
    const db = g.DB!;
    return res.json(db.removeTags(ids, tags));
});

export default router;
